rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\CodeQuality\Rector\Return_;
+
+use PhpParser\Node;
+use Rector\Core\NodeAnalyzer\VariableAnalyzer;
+use Rector\Core\Rector\AbstractRector;
+use PHPStan\Type\ConstantScalarType;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+use PHPStan\Type\Enum\EnumCaseObjectType;
+use PHPStan\Type\VerbosityLevel;
+
+
+/**
+
+ * @see \Rector\Tests\CodeQuality\Rector\Return_\SimplifyReturnExpressionToConstantTypeRector\SimplifyReturnExpressionToConstantTypeRectorTest
+ */
+final class SimplifyReturnExpressionToConstantTypeRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Simplify return expression to constant values', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+class SomeClass
+{
+    public function run()
+    {
+        $return = false;
+
+        if (doSomething()) {
+            return true;
+        }
+
+        return $return;
+    }
+}
+CODE_SAMPLE
+
+                ,
+                <<<'CODE_SAMPLE'
+class SomeClass
+{
+    public function run()
+    {
+        $return = false;
+
+        if (doSomething()) {
+            return true;
+        }
+
+        return false;
+    }
+}
+CODE_SAMPLE
+
+            )
+        ]);
+    }
+
+    public function __construct(
+        private readonly VariableAnalyzer $variableAnalyzer
+    ) {
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [\PhpParser\Node\Stmt\Return_::class];
+    }
+
+    /**
+     * @param \PhpParser\Node\Stmt\Return_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if ($this->shouldSkip($node) || $node->expr === null) {
+            return null;
+        }
+
+        $nativeType = $this->nodeTypeResolver->getNativeType($node->expr);
+        if ($nativeType instanceof ConstantScalarType) {
+            $constantValue = $nativeType->getValue();
```

Comment:
is this work for `++$variable` or `extract()` before return?

I stil think this rule will be buggy, as variable can be modified or replaced magically on the fly before returning.

Reply:
No. Its just the common path. 

I determine whether the type of a return expression is constant, if so return the expression with the constant

Reply:
that need to be added to test fixture imo, then, probably better to be handled via `StmtsAwareInterface` node.

The code like this should be skipped:

1. increment/decrement/postcrement/precrement

```php
$var = 1;

if (rand(0, 1)) {
   return true;
}

++$var;

return $var;
```

2. replaced by `extract()` magically

```php
$var = false;

if (rand(0, 1)) {
   return true;
}

extract($data);

return $var;
```

Reply:
Let me add this tests, thanks

Reply:
added. thx for the cases.

btw, I learned that incrementing `false` stays `false`:
https://3v4l.org/h2NC5